File: Server/src/main/java/cn/rwhps/server/data/plugin/AbstractPluginData.java
/*
 * Copyright 2020-2022 RW-HPS Team and contributors.
 *  
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */

package cn.rwhps.server.data.plugin;

import cn.rwhps.server.func.Prov;
import cn.rwhps.server.io.input.ReusableDisableSyncByteArrayInputStream;
import cn.rwhps.server.struct.ObjectMap;
import cn.rwhps.server.struct.OrderedMap;
import cn.rwhps.server.struct.SerializerTypeAll;
import cn.rwhps.server.util.IsUtil;
import cn.rwhps.server.util.file.FileUtil;
import cn.rwhps.server.util.log.Log;
import cn.rwhps.server.util.log.exp.VariableException;
import cn.rwhps.server.util.zip.gzip.GzipDecoder;
import cn.rwhps.server.util.zip.gzip.GzipEncoder;
import org.jetbrains.annotations.NotNull;

import java.io.*;

/**
 * [PluginData] 的默认实现. 使用 '{@link AbstractPluginData#setData}' 自带创建 [Value] 并跟踪其改动.
 * 实现注意
 * 此类型处于实验性阶段. 使用其中定义的属性和函数是安全的, 但将来可能会新增成员抽象函数.
 * @author RW-HPS/Dr
 */
@SuppressWarnings("unchecked")
class AbstractPluginData {
    private static final ObjectMap<Class<?>, SerializerTypeAll.TypeSerializer<?>> SERIALIZERS = new ObjectMap<>();
    private final OrderedMap<String, Value<?>> PLUGIN_DATA = new OrderedMap<>();
    private final ReusableDisableSyncByteArrayInputStream byteInputStream = new ReusableDisableSyncByteArrayInputStream();
    private final ByteArrayOutputStream byteStream = new ByteArrayOutputStream();
    private final DataOutputStream dataOutput = new DataOutputStream(byteStream);
    private FileUtil fileUtil;

    static {
        DefaultSerializers.register();
    }

    /**
     * 这个 [PluginData] 保存时使用的文件.
     * @param fileUtil FileUtil
     */
    public void setFileUtil(FileUtil fileUtil) {
        this.fileUtil = fileUtil;
        fileUtil.createNewFile();
        this.read();
    }

    /**
     * 向PluginData中加入一个value
     * @param name value的名字
     * @param data 需要存储的数据
     * @param <T> data的类
     */
    public <T> void setData(String name,@NotNull final T data) {
        if (SERIALIZERS.containsKey(data.getClass())) {
            PLUGIN_DATA.put(name, new Value<>(data));
        } else {
            throw new VariableException.ObjectMapRuntimeException("UNSUPPORTED_SERIALIZATION");
        }
    }

    /**
     * 向PluginData中获取一个value
     * @param name value的名字
     * @param <T> data的类
     * @return value
     */
    public <T> T getData(String name) {
        return (T) PLUGIN_DATA.get(name).getData();
    }

    /**
     * 向PluginData中获取一个value
     * @param name value的名字
     * @param data 默认返回的数据
     * @param <T> data的类
     * @return value
     */
    public <T> T getData(String name,@NotNull final T data) {
        return (T) PLUGIN_DATA.get(name, () -> new Value<>(data)).getData();
    }

    public <T> T getData(String name,@NotNull final Prov<T> data) {
        return (T) PLUGIN_DATA.get(name, () -> new Value<>(data.get())).getData();
    }

    public void read() {
        if (IsUtil.isBlank(fileUtil) || fileUtil.notExists() || fileUtil.length() < 1) {
            return;
        }
        try (DataInputStream stream = new DataInputStream(GzipDecoder.getGzipInputStream(fileUtil.getInputsStream()))) {
            int amount = stream.readInt();

            for (int i = 0; i < amount; i++) {
                int length; byte[] bytes;
                String key = stream.readUTF();
                byte type = stream.readByte();
                switch (type) {
                    case 0:
                        PLUGIN_DATA.put(key, new Value<>(stream.readBoolean()));
                        break;
                    case 1:
                        PLUGIN_DATA.put(key, new Value<>(stream.readInt()));
                        break;
                    case 2:
                        PLUGIN_DATA.put(key, new Value<>(stream.readLong()));
                        break;
                    case 3:
                        PLUGIN_DATA.put(key, new Value<>(stream.readFloat()));
                        break;
                    case 4:
                        PLUGIN_DATA.put(key, new Value<>(stream.readUTF()));
                        break;
                    case 5:
                        /* 把String转为Class,来进行反序列化 */
                        Class<?> classCache = Class.forName(stream.readUTF());
                        length = stream.readInt();
                        bytes = new byte[length];
                        stream.read(bytes);
                        PLUGIN_DATA.put(key, new Value<>(getObject(classCache,bytes)));
                        break;
                    default:
                        throw new IllegalStateException("Unexpected value: " + type);
                }
            }
        } catch (EOFException e) {
            // 忽略
            Log.warn("数据文件为空 若第一次启动可忽略");
        } catch (Exception e) {
            Log.error("Read Data",e);
        }
    }

    public void save() {
        if (IsUtil.isBlank(fileUtil) || fileUtil.notExists()) {
            return;
        }
        try(DataOutputStream stream = new DataOutputStream(GzipEncoder.getGzipOutputStream(fileUtil.writeByteOutputStream(false)))){
            stream.writeInt(PLUGIN_DATA.size);

            for(ObjectMap.Entry<String, Value<?>> entry : PLUGIN_DATA.entries()){
                stream.writeUTF(entry.key);
                Object value = entry.value.getData();
                if(value instanceof Boolean){
                    stream.writeByte(0);
                    stream.writeBoolean((Boolean)value);
                }else if(value instanceof Integer){
                    stream.writeByte(1);
                    stream.writeInt((Integer)value);
                }else if(value instanceof Long){
                    stream.writeByte(2);
                    stream.writeLong((Long)value);
                }else if(value instanceof Float){
                    stream.writeByte(3);
                    stream.writeFloat((Float)value);
                }else if(value instanceof String){
                    stream.writeByte(4);
                    stream.writeUTF((String)value);
                }else {
                    try {
                        byte[] bytes = putBytes(value);
                        stream.writeByte(5);
                        /* 去除ToString后的前缀(class com.xxx~) */
                        stream.writeUTF(value.getClass().toString().replace("class ",""));
                        stream.writeInt(bytes.length);
                        stream.write(bytes);
                    } catch (IOException e) {
                        Log.error("Save Error",e);
                    }
                }
            }
            stream.flush();
        }catch(Exception e){
            fileUtil.getFile().delete();
            Log.error("Write Data",e);
            throw new RuntimeException();
        }
    }

    protected static SerializerTypeAll.TypeSerializer getSerializer(Class<?> type) {
        return SERIALIZERS.get(type);
    }

    protected static <T> void setSerializer(Class<?> type, SerializerTypeAll.TypeSerializer<T> ser) {
        SERIALIZERS.put(type, ser);
    }

    protected static <T> void setSerializer(Class<T> type, final SerializerTypeAll.TypeWriter<T> writer, final SerializerTypeAll.TypeReader<T> reader) {
        SERIALIZERS.put(type, new SerializerTypeAll.TypeSerializer<T>() {
            @Override
            public void write(DataOutput stream, T object) throws IOException {
                writer.write(stream, object);
            }
            @Override
            public T read(DataInput stream) throws IOException {
                return reader.read(stream);
            }
        });
    }

    private  <T> T getObject(Class<T> type, byte[] bytes) {
        if (!SERIALIZERS.containsKey(type)) {
            Log.error(new IllegalArgumentException("Type " + type + " does not have a serializer registered!"));
            return null;
        }
        SerializerTypeAll.TypeSerializer<?> serializer = SERIALIZERS.get(type);
        try {
            this.byteInputStream.setBytes(bytes);
            Object obj = serializer.read(new DataInputStream(byteInputStream));
            if (obj == null) {
                return null;
            }
            return (T)obj;
        } catch (Exception e) {
            return null;
        }
    }

    private byte[] putBytes(Object value) throws IOException {
        return putBytes(value, value.getClass());
    }

    private byte[] putBytes(Object value, Class<?> type) throws IOException {
        if (!SERIALIZERS.containsKey(type)) {
            Log.error(new IllegalArgumentException("Type " + type + " does not have a serializer registered!"));
        }
        this.byteStream.reset();
        SerializerTypeAll.TypeSerializer<Object> serializer = (SerializerTypeAll.TypeSerializer)SERIALIZERS.get(type);
        serializer.write(this.dataOutput, value);
        return this.byteStream.toByteArray();
    }
}

File: Server-Core/src/main/java/net/rwhps/server/data/plugin/AbstractPluginData.kt
/*
 * Copyright 2020-2023 RW-HPS Team and contributors.
 *
 * 此源代码的使用受 GNU AFFERO GENERAL PUBLIC LICENSE version 3 许可证的约束, 可以在以下链接找到该许可证.
 * Use of this source code is governed by the GNU AGPLv3 license that can be found through the following link.
 *
 * https://github.com/RW-HPS/RW-HPS/blob/master/LICENSE
 */
package net.rwhps.server.data.plugin

import net.rwhps.server.data.plugin.DefaultSerializers.register
import net.rwhps.server.io.GameInputStream
import net.rwhps.server.io.GameOutputStream
import net.rwhps.server.io.input.ReusableDisableSyncByteArrayInputStream
import net.rwhps.server.io.output.ByteArrayOutputStream
import net.rwhps.server.io.output.CompressOutputStream
import net.rwhps.server.struct.ObjectMap
import net.rwhps.server.struct.OrderedMap
import net.rwhps.server.struct.SerializerTypeAll
import net.rwhps.server.util.IsUtil.isBlank
import net.rwhps.server.util.alone.annotations.NeedToRefactor
import net.rwhps.server.util.compression.CompressionDecoderUtils
import net.rwhps.server.util.compression.gzip.GzipDecoder.getGzipInputStream
import net.rwhps.server.util.file.FileUtil
import net.rwhps.server.util.log.Log.error
import net.rwhps.server.util.log.Log.warn
import net.rwhps.server.util.log.exp.CompressionException
import net.rwhps.server.util.log.exp.VariableException
import java.io.*

/**
 * [PluginData] 的默认实现. 使用 '[AbstractPluginData.setData]' 自带创建 [Value] 并跟踪其改动.
 * 实现注意
 * 此类型处于实验性阶段. 使用其中定义的属性和函数是安全的, 但将来可能会新增成员抽象函数.
 * @author RW-HPS/Dr
 */
@NeedToRefactor
internal open class AbstractPluginData {
    private val pluginData = OrderedMap<String, Value<*>>()
    private val byteInputStream = ReusableDisableSyncByteArrayInputStream()
    private val byteStream = ByteArrayOutputStream()
    private val dataOutput = GameOutputStream(byteStream)
    private var fileUtil: FileUtil? = null
    private var code: String = "gzip"

    /**
     * 这个 [PluginData] 保存时使用的文件.
     * @param fileUtil FileUtil
     */
    fun setFileUtil(fileUtil: FileUtil, code: String = "gzip") {
        this.fileUtil = fileUtil
        fileUtil.createNewFile()
        this.code = code
        this.read()
    }

    /**
     * 向PluginData中加入一个value
     * @param name value的名字
     * @param data 需要存储的数据
     * @param <T> data的类
    </T> */
    inline fun <reified T> setData(name: String, data: T) {
        if (SERIALIZERS.containsKey(T::class.java)) {
            pluginData.put(name, Value(data))
        } else {
            throw VariableException.ObjectMapRuntimeException("${T::class.java} : UNSUPPORTED_SERIALIZATION")
        }
    }

    fun getMap(): OrderedMap<String, Value<*>> {
        return pluginData
    }

    /**
     * 向PluginData中获取一个value
     * @param name value的名字
     * @param <T> data的类
     * @return value
    </T> */
    @Suppress("UNCHECKED_CAST")
    fun <T> getData(name: String): T {
        return pluginData[name].data as T
    }

    /**
     * 向PluginData中获取一个value
     * @param name value的名字
     * @param data 默认返回的数据
     * @param <T> data的类
     * @return value
     */
    @Suppress("UNCHECKED_CAST")
    fun <T> getData(name: String, data: T): T {
        return pluginData[name, { Value(data) }].data as T
    }

    @Suppress("UNCHECKED_CAST")
    fun <T> getData(name: String, data: ()->T): T {
        return pluginData[name, { Value(data()) }].data as T
    }

    fun read() {
        if (isBlank(fileUtil) || fileUtil!!.notExists() || fileUtil!!.length() < 1) {
            return
        }
        try {
            fileUtil!!.getInputsStream().use { stream -> read(stream) }
        } catch (e: Exception) {
            error("[Read BIN Error]",e)
        }
    }

    @Throws(CompressionException.CryptographicException::class)
    fun read(inStream: InputStream) {
        val gameInputStream: GameInputStream =
            when(code) {
                "7z" -> GameInputStream(CompressionDecoderUtils.lz77Stream(inStream).getZipAllBytes()["file"])
                "gzip" -> GameInputStream(getGzipInputStream(inStream))
                else -> throw CompressionException.CryptographicException(code)
            }

        try {
            gameInputStream.use { stream ->
                val amount = stream.readInt()
                for (i in 0 until amount) {
                    var length: Int
                    var bytes: ByteArray
                    val key = stream.readString()
                    when (val type = stream.readByte()) {
                        0 -> pluginData.put(key, Value(stream.readBoolean()))
                        1 -> pluginData.put(key, Value(stream.readInt()))
                        2 -> pluginData.put(key, Value(stream.readLong()))
                        3 -> pluginData.put(key, Value(stream.readFloat()))
                        4 -> pluginData.put(key, Value(stream.readString()))
                        5 -> {
                            /* 把String转为Class,来进行反序列化 */
                            val classCache: Class<*> = Class.forName(stream.readString().replace("net.rwhps.server", "net.rwhps.server"))
                            length = stream.readInt()
                            bytes = stream.readNBytes(length)
                            pluginData.put(key, Value(getObject(classCache, bytes)))
                        }

                        else -> throw IllegalStateException("Unexpected value: $type")
                    }
                }
            }
        } catch (e: EOFException) {
            // 忽略
            warn("数据文件为空 若第一次启动可忽略")
        } catch (e: Exception) {
            error("Read Data", e)
        }
    }

    fun save() {
        if (isBlank(fileUtil) || fileUtil!!.notExists()) {
            return
        }
        try {
            fileUtil!!.writeByteOutputStream(false).use { stream -> save(stream) }
        } catch (e: Exception) {
            error("[Write BIN Error]",e)
        }
    }

    fun save(outputStream: OutputStream) {
        val gameOutputStream: GameOutputStream =
            when(code) {
                "7z" -> CompressOutputStream.get7zOutputStream("",true)
                "gzip" -> CompressOutputStream.getGzipOutputStream("",true)
                else -> throw CompressionException.CryptographicException(code)
            }

        try {
            gameOutputStream.use { stream ->
                stream.writeInt(pluginData.size)
                for (entry in pluginData) {
                    stream.writeString(entry.key)
                    when (val value = entry.value.data!!) {
                        is Boolean -> {
                            stream.writeByte(0)
                            stream.writeBoolean(value)
                        }
                        is Int -> {
                            stream.writeByte(1)
                            stream.writeInt(value)
                        }
                        is Long -> {
                            stream.writeByte(2)
                            stream.writeLong(value)
                        }
                        is Float -> {
                            stream.writeByte(3)
                            stream.writeFloat(value)
                        }
                        is String -> {
                            stream.writeByte(4)
                            stream.writeString(value)
                        }
                        else -> {
                            try {
                                val bytes = putBytes(value)
                                stream.writeByte(5)
                                /* 去除ToString后的前缀(class com.xxx~) */
                                stream.writeString(value.javaClass.toString().replace("class ", ""))
                                stream.writeInt(bytes.size)
                                stream.writeBytes(bytes)
                            } catch (e: IOException) {
                                error("Save Error", e)
                            }
                        }
                    }
                }
            }
            outputStream.write(gameOutputStream.getByteArray())
            outputStream.flush()
        } catch (e: Exception) {
            fileUtil!!.file.delete()
            error("Write Data", e)
            throw RuntimeException()
        }
    }

    fun cleanRam() {
        pluginData.clear()
    }

    private fun <T> getObject(type: Class<T>, bytes: ByteArray): T? {
        if (!SERIALIZERS.containsKey(type)) {
            error(IllegalArgumentException("Type $type does not have a serializer registered!"))
            return null
        }
        val serializer = SERIALIZERS[type]
        return try {
            byteInputStream.setBytes(bytes)
            val obj = serializer.read(GameInputStream(byteInputStream)) ?: return null
            @Suppress("UNCHECKED_CAST")
            obj as T
        } catch (e: Exception) {
            null
        }
    }

    @Throws(IOException::class)
    private fun putBytes(value: Any, type: Class<*> = value.javaClass): ByteArray {
        if (!SERIALIZERS.containsKey(type)) {
            error(IllegalArgumentException("Type $type does not have a serializer registered!"))
        }
        byteStream.reset()
        val serializer: SerializerTypeAll.TypeSerializer<Any?> = SERIALIZERS[type] as SerializerTypeAll.TypeSerializer<in Any?>
        serializer.write(dataOutput, value)
        return byteStream.toByteArray()
    }

    companion object {
        private val SERIALIZERS = ObjectMap<Class<*>, SerializerTypeAll.TypeSerializer<Any?>>()

        init {
            register()
        }

        internal fun getSerializer(type: Class<*>): SerializerTypeAll.TypeSerializer<Any?>? {
            return SERIALIZERS[type]
        }

        internal fun <T> setSerializer(type: Class<*>, ser: SerializerTypeAll.TypeSerializer<T>) {
            @Suppress("UNCHECKED_CAST")
            SERIALIZERS.put(type, ser as SerializerTypeAll.TypeSerializer<Any?>)
        }
    }
}